File: problems/1/1.two-sum.py
#
# @lc app=leetcode id=1 lang=python3
#
# [1] Two Sum
#

# @lc code=start
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # find pair of integers that add up to target
        # nums = [2, 7, 11, 5], target = 9
        # for each val, store in hashmap (target - nums[i] -> i)

        seen = {}
        for i, val in enumerate(nums):
            if val in seen:
                return [seen[val], i]
            seen[target - val] = i

        
# @lc code=end

File: problems/1/1.two-sum.rb
#
# @lc app=leetcode id=1 lang=ruby
#
# [1] Two Sum
#

# @lc code=start
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def two_sum(nums, target)
  cache = {}
  nums.each_with_index do |num, i|
    return [cache[num], i] if cache.include?(num)

    cache[target - num] = i
  end
end
# @lc code=end
